webscale/tls/aliaskeyedcache.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2021 Webscale
+
+package tls
+
+import (
+	"crypto/tls"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/mholt/certmagic"
+)
+
+// Unique type for strings destined to be Caddy instance keys.
+type instanceKey string
+
+// The override cache instance key.
+const OverrideCacheKey instanceKey = "strict-tls"
+
+// AliasKeyedCache is a certificate cache that serves certificates based solely
+// on the SNI in the client hello message during TLS handshakes. As such, it
+// ignores what domains are covered by the certificate it is serving, checking
+// only the SNI. The alias-certificate associations are written in the
+// Caddyfile's 'strict_tls' directive. If a certificate is not found based on
+// the SNI, the default certificate, configured from 'strict_tls.default' is
+// served.
+type AliasKeyedCache struct {
+	// The cache is keyed by hostname.
+	cache map[string]*certmagic.Certificate
+
+	// Default certificate.
+	defaultCertificate *certmagic.Certificate
+
+	// Protects the cache map.
+	mu sync.RWMutex
+}
+
+// NewAliasKeyedCache creates a new alias-certificate cache.
+// Only Webscale's 'strict_tls' plugin should call this function.
+func NewAliasKeyedCache() *AliasKeyedCache {
+	return &AliasKeyedCache{
+		cache:              make(map[string]*certmagic.Certificate),
+		defaultCertificate: nil,
+	}
+}
+
+// Load adds new alias-certificate associations for the given *.pem file and the
+// given list of aliases. Each alias will be associated to the given
+// certificate. If the given file is not a valid PEM file, an error is returned.
+// Only Webscale's 'strict_tls' plugin should call this function.
+func (cache *AliasKeyedCache) Load(file string, aliases []string) error {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
```

Comment:
The mutex lock is only needed right before the loop on line 65 below, right? Reading the file from the disk and decrypting the contents within may not be trivial in terms of time, so should we move the lock until after the parsing/making the certificate is complete?

Reply:
On second thoughts, as we discussed, do we need a mutex protection here? Seems like the map will be unique to each Caddy instance that gets spawned at reload. And at that time no requests are being served by that instance, so no one should be reading from the map even. The older instance would be serving requests and reading from the cache, but that is an independent copy, so I am not sure if this map will ever be accessed by multiple threads.

Reply:
I agree. I can't think of an instance where an active Caddy generation would modify the certificate cache after initiializing from the Caddyfile.